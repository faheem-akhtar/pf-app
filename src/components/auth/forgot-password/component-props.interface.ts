export interface AuthForgotPasswordComponentPropsInterface {
  /**
   * On close auth modal
   */
  onClose: () => void;

  /**
   * On success
   */
  onSuccess: () => void;

  /**
   * On registration link clicked
   */
  onRegister: () => void;
}
