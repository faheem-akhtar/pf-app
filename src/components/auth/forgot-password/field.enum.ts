export enum AuthForgotPasswordFieldEnum {
  /**
   * Email
   */
  email = 'email',
}
