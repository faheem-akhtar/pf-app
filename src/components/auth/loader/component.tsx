import { domClassMerge } from 'helpers/dom/class-merge';

import styles from './auth-loader.module.scss';

export const AuthLoaderComponent = ({
  isEnabled,
  isCentered,
}: {
  isEnabled: boolean;
  isCentered: boolean;
}): JSX.Element | null =>
  isEnabled ? (
    <div className={styles.loader1}>
      <div
        className={domClassMerge(styles['loader1__spinner'], isCentered ? styles['loader1__spinner--centered'] : '')}
      >
        <div className={domClassMerge(styles['loader1__bounce'], styles['loader1__bounce--1'])} />
        <div className={domClassMerge(styles['loader1__bounce'], styles['loader1__bounce--2'])} />
        <div className={domClassMerge(styles['loader1__bounce'], styles['loader1__bounce--3'])} />
      </div>
    </div>
  ) : null;
