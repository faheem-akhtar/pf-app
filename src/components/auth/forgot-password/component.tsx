import React, { useState } from 'react';

import { AuthForgotPasswordFieldEnum } from 'components/auth/forgot-password/field.enum';
import { AuthForgotPasswordPropsInterface } from 'components/auth/forgot-password/props.interface';
import { AuthLoaderComponent } from 'components/auth/loader/component';
import { AuthLoginFieldEnum } from 'components/auth/login/field.enum';
import { AuthResetPasswordService } from 'services/auth/reset-password.service';
import { ButtonComponentTypeEnum } from 'library/button/component-type.enum';
import { ButtonSizeEnum } from 'library/button/size.enum';
import { ButtonTemplate } from 'library/button/template';
import { domClassMerge } from 'helpers/dom/class-merge';
import { ErrorMessageComponent } from 'components/error-message/component';
import { formMakeValidator } from 'components/form/make-validator';
import { GoogleRecaptchaService } from 'services/google/recaptcha.service';
import { InputBaseComponent } from 'library/input/base/component';
import { ReCaptchaComponent } from 'components/re-captcha/component';
import { useTranslation } from 'helpers/translation/hook';
import { validationEmail } from 'helpers/validation/email';
import { validationRequired } from 'helpers/validation/required';

import styles from '../auth.module.scss';

export const AuthForgotPasswordComponent = (props: AuthForgotPasswordPropsInterface): JSX.Element => {
  const captchaService = GoogleRecaptchaService();
  const { t } = useTranslation();
  const [email, setEmail] = useState('');
  const [errors, setErrors] = useState<Partial<Record<AuthLoginFieldEnum, string>>>({});
  const [errorMessage, setErrorMessage] = useState('');
  const [validators] = useState(() => ({
    [AuthLoginFieldEnum.email]: [
      validationRequired(t('validation/empty-email')),
      validationEmail(t('validation/not-valid-email')),
    ],
  }));
  const [isLoading, setIsLoading] = useState(false);

  /**
   * Validate form fields
   */
  const validate = formMakeValidator(errors, setErrors, validators);

  // TODO-FE[CX-499] - Add success message
  return (
    <div data-testid='AuthForgotPasswordComponent'>
      <div className={styles.heading}>{t('auth/forgot-password')}?</div>
      <AuthLoaderComponent isEnabled={isLoading} isCentered={true} />
      {errorMessage && <ErrorMessageComponent message={errorMessage} />}
      <form
        onSubmit={(e): void => {
          e.preventDefault();

          if (
            !validate({
              [AuthForgotPasswordFieldEnum.email]: email,
            })
          ) {
            setIsLoading(true);
            captchaService.execute().then((captcha_token) => {
              AuthResetPasswordService({
                email,
                captcha_token,
              })
                .then((e) => {
                  if (e.ok) {
                    // Reset server error message
                    setErrorMessage('');
                    // Close modal
                    props.onClose();
                    props.onSuccess();
                  } else {
                    setErrorMessage(e.error?.body || `${t('auth/something-wrong')}! ${t('auth/try-later')}`);
                    captchaService.reset();
                  }
                })
                .finally(() => setIsLoading(false));
            });
          }
          return;
        }}
      >
        <div className={styles['input-area']}>
          <InputBaseComponent
            placeholder={t('email')}
            type='email'
            value={email}
            error={!!errors[AuthForgotPasswordFieldEnum.email]}
            onChange={(value): void => {
              validate({
                [AuthForgotPasswordFieldEnum.email]: value,
              });
              setEmail(value);
            }}
            errorText={errors[AuthForgotPasswordFieldEnum.email]}
          />
        </div>
        <div className={styles['input-area']}>
          <ButtonTemplate
            className={styles.button}
            type='submit'
            componentType={ButtonComponentTypeEnum.primary}
            size={ButtonSizeEnum.regular}
          >
            {t('auth/reset-password')}
          </ButtonTemplate>
        </div>
        <ReCaptchaComponent />
      </form>
      <div className={domClassMerge(styles.link, styles['create-account'])} onClick={props.onRegister}>
        {t('auth/not-registered-yet')}?&nbsp;{t('auth/create-account')}
      </div>
    </div>
  );
};
