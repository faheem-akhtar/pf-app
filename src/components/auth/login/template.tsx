import React, { FunctionComponent } from 'react';

import { AuthLoaderComponent } from 'components/auth/loader/component';
import { AuthLoginFieldEnum } from 'components/auth/login/field.enum';
import { ErrorMessageComponent } from 'components/error-message/component';
import { IconFacebookTemplate } from 'components/icon/facebook-template';
import { IconGoogleTemplate } from 'components/icon/google-template';
import { ReCaptchaComponent } from 'components/re-captcha/component';
import { domClassMerge } from 'helpers/dom/class-merge';
import { ButtonComponentTypeEnum } from 'library/button/component-type.enum';
import { ButtonIconPositionEnum } from 'library/button/icon-position.enum';
import { ButtonSizeEnum } from 'library/button/size.enum';
import { ButtonTemplate } from 'library/button/template';
import { InputBaseComponent } from 'library/input/base/component';

import styles from '../auth.module.scss';
import { AuthLoginTemplatePropsInterface } from './template-props.interface';

export const AuthLoginTemplate: FunctionComponent<AuthLoginTemplatePropsInterface> = (props) => (
  <div data-testid='AuthLoginComponent'>
    <h1 className={styles.heading}>{props.t('sign-in')}</h1>
    {/* Facebook */}
    <ButtonTemplate
      type='button'
      className={domClassMerge(styles.button, styles.facebook)}
      componentType={ButtonComponentTypeEnum.tertiary}
      size={ButtonSizeEnum.small}
      onClick={props.onFacebookLoginClick}
      icon={{ component: IconFacebookTemplate, position: ButtonIconPositionEnum.left }}
    >
      {props.t('auth/sign-in-facebook')}
    </ButtonTemplate>
    {/* Google */}
    <ButtonTemplate
      type='button'
      className={domClassMerge(styles.button, styles.google)}
      componentType={ButtonComponentTypeEnum.tertiary}
      size={ButtonSizeEnum.small}
      onClick={props.onGoogleLoginClick}
      icon={{ component: IconGoogleTemplate, position: ButtonIconPositionEnum.left }}
    >
      {props.t('auth/sign-in-google')}
    </ButtonTemplate>
    <div className={styles.divider}>
      <span className={styles.text}>{props.t('auth/or')}</span>
    </div>
    <AuthLoaderComponent isEnabled={props.isLoading} isCentered={true} />
    {props.errorMessage && <ErrorMessageComponent message={props.errorMessage} />}
    <form onSubmit={props.onFormSubmit}>
      <div className={styles['input-area']}>
        <InputBaseComponent
          id='email'
          name='email'
          placeholder={props.t('email')}
          type='email'
          value={props.email}
          error={!!props.errors[AuthLoginFieldEnum.email]}
          onChange={props.onEmailChange}
          errorText={props.errors[AuthLoginFieldEnum.email]}
        />
      </div>
      <div className={styles['input-area']}>
        <InputBaseComponent
          id='password'
          name='password'
          placeholder={props.t('password')}
          type='password'
          value={props.password}
          error={!!props.errors[AuthLoginFieldEnum.password]}
          onChange={props.onPasswordChange}
          errorText={props.errors[AuthLoginFieldEnum.password]}
        />
      </div>
      <div className={styles['input-area']} onClick={props.onForgotPassword}>
        <div className={domClassMerge(styles.link, styles['forgot-password'])}>{props.t('auth/forgot-password')}?</div>
      </div>
      <div className={styles['input-area']}>
        <ButtonTemplate
          className={styles.button}
          type='submit'
          componentType={ButtonComponentTypeEnum.primary}
          size={ButtonSizeEnum.regular}
        >
          {props.t('log-in')}
        </ButtonTemplate>
      </div>
      <ReCaptchaComponent />
    </form>
    <div className={domClassMerge(styles.link, styles['create-account'])} onClick={props.onRegister}>
      {props.t('auth/not-registered-yet')}?&nbsp;{props.t('auth/create-account')}
    </div>
  </div>
);
