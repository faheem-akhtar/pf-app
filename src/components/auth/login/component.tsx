import React, { useState } from 'react';

import { AuthFacebookService } from 'services/auth/facebook.service';
import { AuthGoogleService } from 'services/auth/google.service';
import { AuthLoaderComponent } from 'components/auth/loader/component';
import { AuthLoginFieldEnum } from 'components/auth/login/field.enum';
import { AuthLoginPropsInterface } from 'components/auth/login/props.interface';
import { AuthLoginService } from 'services/auth/login.service';
import { ButtonComponentTypeEnum } from 'library/button/component-type.enum';
import { ButtonIconPositionEnum } from 'library/button/icon-position.enum';
import { ButtonSizeEnum } from 'library/button/size.enum';
import { ButtonTemplate } from 'library/button/template';
import { domClassMerge } from 'helpers/dom/class-merge';
import { ErrorMessageComponent } from 'components/error-message/component';
import { formMakeValidator } from 'components/form/make-validator';
import { GoogleRecaptchaService } from 'services/google-recaptcha/service';
import { IconFacebookTemplate } from 'components/icon/facebook-template';
import { IconGoogleTemplate } from 'components/icon/google-template';
import { InputBaseComponent } from 'library/input/base/component';
import { ReCaptchaComponent } from 'components/re-captcha/component';
import { useTranslation } from 'helpers/translation/hook';
import { validationEmail } from 'helpers/validation/email';
import { validationRequired } from 'helpers/validation/required';

import styles from '../auth.module.scss';
const captchaService = new GoogleRecaptchaService();

export const AuthLoginComponent = (props: AuthLoginPropsInterface): JSX.Element => {
  const { t } = useTranslation();

  const [email, setEmail] = useState('');
  const [password, setPassword] = useState('');
  const [errors, setErrors] = useState<Partial<Record<AuthLoginFieldEnum, string>>>({});
  const [errorMessage, setErrorMessage] = useState('');
  const [validators] = useState(() => ({
    [AuthLoginFieldEnum.email]: [
      validationRequired(t('validation/empty-email')),
      validationEmail(t('validation/not-valid-email')),
    ],
    [AuthLoginFieldEnum.password]: [validationRequired(t('auth/empty-password'))],
  }));
  const [isLoading, setIsLoading] = useState(false);

  /**
   * Validate form fields
   */
  const validate = formMakeValidator(errors, setErrors, validators);

  return (
    <div>
      <div className={styles.heading}>{t('sign-in')}</div>
      {/* Facebook */}
      <ButtonTemplate
        type='button'
        className={domClassMerge(styles.button, styles.facebook)}
        componentType={ButtonComponentTypeEnum.tertiary}
        size={ButtonSizeEnum.small}
        onClick={(): void => {
          setIsLoading(true);
          AuthFacebookService.signIn()
            .then(props.onClose)
            .catch(() => {
              setErrorMessage(`${t('auth/something-wrong')}! ${t('auth/try-later')}`);
              props.onClose();
            })
            .finally(() => setIsLoading(false));
        }}
        icon={{ component: IconFacebookTemplate, position: ButtonIconPositionEnum.left }}
      >
        {t('auth/sign-in-facebook')}
      </ButtonTemplate>
      {/* Google */}
      <ButtonTemplate
        type='button'
        className={domClassMerge(styles.button, styles.google)}
        componentType={ButtonComponentTypeEnum.tertiary}
        size={ButtonSizeEnum.small}
        onClick={(): void => {
          setIsLoading(true);
          AuthGoogleService.signIn()
            .then(props.onClose)
            .catch(() => {
              setErrorMessage(`${t('auth/something-wrong')}! ${t('auth/try-later')}`);
              props.onClose();
            })
            .finally(() => setIsLoading(false));
        }}
        icon={{ component: IconGoogleTemplate, position: ButtonIconPositionEnum.left }}
      >
        {t('auth/sign-in-google')}
      </ButtonTemplate>
      <div className={styles.divider}>
        <span className={styles.text}>{t('auth/or')}</span>
      </div>
      <AuthLoaderComponent isEnabled={isLoading} isCentered={true} />
      {errorMessage && <ErrorMessageComponent message={errorMessage} />}
      <form
        onSubmit={(e): void => {
          e.preventDefault();

          if (
            !validate({
              [AuthLoginFieldEnum.email]: email,
              [AuthLoginFieldEnum.password]: password,
            })
          ) {
            setIsLoading(true);
            captchaService.execute().then((captcha_token) => {
              AuthLoginService({
                email,
                password,
                captcha_token,
              })
                .then((e) => {
                  if (e.ok) {
                    // Reset server error message
                    setErrorMessage('');
                    // Close modal
                    props.onClose();
                  } else {
                    setErrorMessage(e.error.body || `${t('auth/something-wrong')}! ${t('auth/try-later')}`);
                    captchaService.reset();
                  }
                })
                .finally(() => setIsLoading(false));
            });
          }
          return;
        }}
      >
        <div className={styles['input-area']}>
          <InputBaseComponent
            placeholder={t('email')}
            type='email'
            value={email}
            error={!!errors[AuthLoginFieldEnum.email]}
            onChange={(value): void => {
              validate({
                [AuthLoginFieldEnum.email]: value,
              });
              setEmail(value);
            }}
            errorText={errors[AuthLoginFieldEnum.email]}
          />
        </div>
        <div className={styles['input-area']}>
          <InputBaseComponent
            placeholder={t('password')}
            type='password'
            value={password}
            error={!!errors[AuthLoginFieldEnum.password]}
            onChange={(value): void => {
              !validate({
                [AuthLoginFieldEnum.password]: value,
              });
              setPassword(value);
            }}
            errorText={errors[AuthLoginFieldEnum.password]}
          />
        </div>
        <div className={styles['input-area']} onClick={props.onForgotPassword}>
          <div className={domClassMerge(styles.link, styles['forgot-password'])}>{t('auth/forgot-password')}?</div>
        </div>
        <div className={styles['input-area']}>
          <ButtonTemplate
            className={styles.button}
            type='submit'
            componentType={ButtonComponentTypeEnum.primary}
            size={ButtonSizeEnum.regular}
          >
            {t('log-in')}
          </ButtonTemplate>
        </div>
        <ReCaptchaComponent />
      </form>
      <div className={domClassMerge(styles.link, styles['create-account'])} onClick={props.onRegister}>
        {t('auth/not-registered-yet')}?&nbsp;{t('auth/create-account')}
      </div>
    </div>
  );
};
