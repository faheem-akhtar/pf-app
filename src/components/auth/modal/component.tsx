import React, { useState } from 'react';

import { AuthForgotPasswordComponent } from 'components/auth/forgot-password/component';
import { AuthLoginComponent } from 'components/auth/login/component';
import { AuthRegistrationComponent } from 'components/auth/registration/component';
import { AuthScreenEnum } from 'components/auth/screen.enum';
import { IconThickCrossTemplate } from 'components/icon/thick/cross-template';
import { functionNoop } from 'helpers/function/noop';

import styles from './auth-modal-component.module.scss';
import { AuthModalPropsInterface } from './props.interface';

export const AuthModalComponent = ({
  close = functionNoop,
  cancel = functionNoop,
  success = functionNoop,
}: AuthModalPropsInterface): JSX.Element => {
  const [authScreen, setAuthScreen] = useState(AuthScreenEnum.login);
  return (
    <div
      className={styles.content}
      onClick={(e): void => {
        e.stopPropagation();
      }}
    >
      <div className={styles.header}>
        <button
          onClick={(): void => {
            close();
            cancel();
          }}
        >
          <IconThickCrossTemplate class={styles.closeIcon} />
        </button>
      </div>
      {authScreen === AuthScreenEnum.login && (
        <AuthLoginComponent
          onClose={close}
          onSuccess={success}
          onRegister={(): void => setAuthScreen(AuthScreenEnum.registration)}
          onForgotPassword={(): void => setAuthScreen(AuthScreenEnum.forgotPassword)}
        />
      )}
      {authScreen === AuthScreenEnum.registration && (
        <AuthRegistrationComponent
          onClose={close}
          onSuccess={success}
          onLogin={(): void => setAuthScreen(AuthScreenEnum.login)}
        />
      )}
      {authScreen === AuthScreenEnum.forgotPassword && (
        <AuthForgotPasswordComponent
          onClose={close}
          onSuccess={success}
          onRegister={(): void => setAuthScreen(AuthScreenEnum.registration)}
        />
      )}
    </div>
  );
};
