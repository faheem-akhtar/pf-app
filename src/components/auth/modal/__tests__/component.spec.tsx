/**
 * @jest-environment jsdom
 */
import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import * as AuthForgotPasswordComponentModule from 'components/auth/forgot-password/component';
import * as AuthLoginComponentModule from 'components/auth/login/component';
import * as AuthRegistrationComponentModule from 'components/auth/registration/component';
import { AuthScreenEnum } from 'components/auth/screen.enum';
import { AuthSuccessTypeEnum } from 'components/auth/success-type.enum';
import { SavedPropertyAuthLoginTemplate } from 'components/saved-property/auth/login/template';

import { AuthModalComponent } from '../component';
import { AuthModalPropsInterface } from '../props.interface';

describe('AuthModalComponent', () => {
  let props: AuthModalPropsInterface;

  beforeEach(() => {
    window.dataLayer = [];
    props = {
      close: jest.fn(),
      cancel: jest.fn(),
      success: jest.fn(),
    };
  });

  it('renders without throwing any errors', () => {
    expect(render(<AuthModalComponent {...props} />).container).toMatchSnapshot();
  });

  it('should accept an event label', () => {
    render(<AuthModalComponent {...props} eventLabel='event label' />);

    expect(window.dataLayer).toEqual([
      {
        event: 'Userbox',
        eventAction: 'Click:Sign in Button',
        eventLabel: 'event label',
      },
    ]);
  });

  describe('login screen', () => {
    beforeEach(() => {
      render(<AuthModalComponent {...props} />);

      window.dataLayer = [];
    });

    it('should render login by default', () => {
      expect(screen.getByRole('heading', { level: 1, name: 'sign-in' })).toBeInTheDocument();
    });

    it('should display registration screen', async () => {
      userEvent.click(screen.getByText('auth/not-registered-yet', { exact: false }));

      expect(screen.getByRole('heading', { level: 1, name: 'auth/create-account' })).toBeInTheDocument();

      await waitFor(() => {
        expect(window.dataLayer).toEqual([
          {
            event: 'Userbox',
            eventAction: 'Start:Sign up with Email',
            eventLabel: 'Header',
          },
        ]);
      });
    });

    it('should display forgot password screen', async () => {
      userEvent.click(screen.getByText('auth/forgot-password?'));
      expect(screen.getByRole('heading', { level: 1, name: 'auth/forgot-password?' })).toBeInTheDocument();

      await waitFor(() => {
        expect(window.dataLayer).toEqual([
          {
            event: 'Userbox',
            eventAction: 'Start:Password Reset',
            eventLabel: 'Header',
          },
        ]);
      });
    });

    it('should call close', () => {
      jest
        .spyOn(AuthLoginComponentModule, 'AuthLoginComponent')
        .mockImplementationOnce(({ onClose }) => <button data-testid='close-button' onClick={onClose} />);
      render(<AuthModalComponent {...props} />);

      userEvent.click(screen.getByTestId('close-button'));
      expect(props.close).toHaveBeenCalledTimes(1);
    });

    it('should call close on header', () => {
      userEvent.click(screen.getByRole('button', { name: /cross/i }));
      expect(props.close).toHaveBeenCalledTimes(1);
    });

    it('should call success', () => {
      jest
        .spyOn(AuthLoginComponentModule, 'AuthLoginComponent')
        .mockImplementationOnce(({ onSuccess }) => (
          <button data-testid='my-button' onClick={(): void => onSuccess(AuthSuccessTypeEnum.signInWithEmail)} />
        ));
      render(<AuthModalComponent {...props} />);

      window.dataLayer = [];

      userEvent.click(screen.getByTestId('my-button'));

      expect(props.success).toHaveBeenCalledTimes(1);

      expect(window.dataLayer).toEqual([
        {
          event: 'Userbox',
          eventAction: 'Finish:Sign in with email',
          eventLabel: 'Header',
        },
      ]);
    });

    it('should track success sign in with facebook', () => {
      jest
        .spyOn(AuthLoginComponentModule, 'AuthLoginComponent')
        .mockImplementationOnce(({ onSuccess }) => (
          <button data-testid='my-button' onClick={(): void => onSuccess(AuthSuccessTypeEnum.signInWithFacebook)} />
        ));
      render(<AuthModalComponent {...props} />);

      window.dataLayer = [];

      userEvent.click(screen.getByTestId('my-button'));

      expect(window.dataLayer).toEqual([
        {
          event: 'Userbox',
          eventAction: 'Finish:Sign in with facebook',
          eventLabel: 'Header',
        },
      ]);
    });

    it('should track success sign in with google', () => {
      jest
        .spyOn(AuthLoginComponentModule, 'AuthLoginComponent')
        .mockImplementationOnce(({ onSuccess }) => (
          <button data-testid='my-button' onClick={(): void => onSuccess(AuthSuccessTypeEnum.signInWithGoogle)} />
        ));
      render(<AuthModalComponent {...props} />);

      window.dataLayer = [];

      userEvent.click(screen.getByTestId('my-button'));

      expect(window.dataLayer).toEqual([
        {
          event: 'Userbox',
          eventAction: 'Finish:Sign in with google',
          eventLabel: 'Header',
        },
      ]);
    });
  });

  describe('short login screen', () => {
    beforeEach(() => {
      render(
        <AuthModalComponent
          {...props}
          initialScreen={AuthScreenEnum.shortLogin}
          loginTemplate={SavedPropertyAuthLoginTemplate}
        />
      );
    });

    it('should render the short login screen', () => {
      expect(screen.getByRole('heading', { level: 1, name: 'saved_property/auth-login-title' })).toBeInTheDocument();
    });

    it('should switch to default login screen', () => {
      userEvent.click(screen.getByText('log-in'));

      expect(screen.getByRole('heading', { level: 1, name: 'sign-in' })).toBeInTheDocument();
    });

    ['google', 'facebook'].forEach((platform) => {
      it(`should track "start sign in with ${platform}" event`, () => {
        userEvent.click(screen.getByRole('button', { name: `auth/sign-in-${platform}` }));

        expect(window.dataLayer).toEqual([
          {
            event: 'Userbox',
            eventAction: `Start:Sign in with ${platform}`,
            eventLabel: 'Header',
          },
        ]);
      });
    });
  });

  describe('registration screen', () => {
    it('should switch to login screen', () => {
      render(<AuthModalComponent {...props} initialScreen={AuthScreenEnum.registration} />);
      userEvent.click(screen.getByText('auth/already-registered? log-in'));

      expect(screen.getByRole('heading', { level: 1, name: 'sign-in' })).toBeInTheDocument();
    });

    it('should call close', () => {
      jest
        .spyOn(AuthRegistrationComponentModule, 'AuthRegistrationComponent')
        .mockImplementationOnce(({ onClose }) => <button data-testid='my-button' onClick={onClose} />);

      render(<AuthModalComponent {...props} initialScreen={AuthScreenEnum.registration} />);

      userEvent.click(screen.getByTestId('my-button'));
      expect(props.close).toHaveBeenCalledTimes(1);
    });

    it('should call success', () => {
      jest
        .spyOn(AuthRegistrationComponentModule, 'AuthRegistrationComponent')
        .mockImplementationOnce(({ onSuccess }) => (
          <button data-testid='my-button' onClick={(): void => onSuccess(AuthSuccessTypeEnum.registerWithEmail)} />
        ));
      render(<AuthModalComponent {...props} initialScreen={AuthScreenEnum.registration} />);

      userEvent.click(screen.getByTestId('my-button'));
      expect(props.success).toHaveBeenCalledTimes(1);
    });
  });

  describe('forgot password screen', () => {
    it('should switch to registration screen', () => {
      render(<AuthModalComponent {...props} initialScreen={AuthScreenEnum.forgotPassword} />);

      userEvent.click(screen.getByText('auth/not-registered-yet? auth/create-account'));

      expect(screen.getByRole('heading', { level: 1, name: 'auth/create-account' })).toBeInTheDocument();
    });

    it('should call close', () => {
      jest
        .spyOn(AuthForgotPasswordComponentModule, 'AuthForgotPasswordComponent')
        .mockImplementationOnce(({ onClose }) => <button data-testid='my-button' onClick={onClose} />);

      render(<AuthModalComponent {...props} initialScreen={AuthScreenEnum.forgotPassword} />);

      userEvent.click(screen.getByTestId('my-button'));
      expect(props.close).toHaveBeenCalledTimes(1);
    });

    it('should call success', () => {
      jest
        .spyOn(AuthForgotPasswordComponentModule, 'AuthForgotPasswordComponent')
        .mockImplementationOnce(({ onSuccess }) => <button data-testid='my-button' onClick={onSuccess} />);

      render(<AuthModalComponent {...props} initialScreen={AuthScreenEnum.forgotPassword} />);

      userEvent.click(screen.getByTestId('my-button'));
      expect(props.success).toHaveBeenCalledTimes(1);
    });
  });
});
