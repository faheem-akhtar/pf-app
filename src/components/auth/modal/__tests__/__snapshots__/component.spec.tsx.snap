// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthModalComponent renders without throwing any errors 1`] = `
<div>
  <div
    class="content"
  >
    <div
      class="header"
    >
      <button>
        <svg
          class="closeIcon"
          viewBox="7 7 10 10"
        >
          <title>
            Cross
          </title>
          <path
            clip-rule="evenodd"
            d="M8.34408 7.17475C8.01717 6.92057 7.54448 6.94367 7.24408 7.24408C6.91864 7.56951 6.91864 8.09715 7.24408 8.42259L10.8215 12L7.24408 15.5774C6.91864 15.9028 6.91864 16.4305 7.24408 16.7559C7.54448 17.0563 8.01717 17.0794 8.34408 16.8252L8.42259 16.7559L12 13.1785L15.5774 16.7559L15.6559 16.8252C15.9828 17.0794 16.4555 17.0563 16.7559 16.7559C17.0814 16.4305 17.0814 15.9028 16.7559 15.5774L13.1785 12L16.7559 8.42259C17.0814 8.09715 17.0814 7.56951 16.7559 7.24408C16.4555 6.94367 15.9828 6.92057 15.6559 7.17475L15.5774 7.24408L12 10.8215L8.42259 7.24408L8.34408 7.17475Z"
            fill-rule="evenodd"
          />
        </svg>
      </button>
    </div>
    <div>
      <h1
        class="heading"
      >
        sign-in
      </h1>
      <button
        class="button tertiary small button facebook"
      >
        <svg
          class="icon icon--left"
          viewBox="0 0 12 23"
        >
          <g
            fill="none"
            fill-rule="evenodd"
            stroke="none"
            stroke-width="1"
          >
            <g
              fill="#FEFEFE"
              transform="translate(-190.000000, -198.000000)"
            >
              <g
                transform="translate(166.000000, 186.000000)"
              >
                <g>
                  <path
                    d="M31.9328743,35 L27.6347873,35 L27.6347873,23.5309121 L24.7675854,23.5309121 L24.7675854,19.5782401 L27.6347873,19.5768961 L27.6301953,17.248481 C27.6301953,14.0238238 28.5045409,12.0618241 32.3028177,12.0618241 L35.4653137,12.0618241 L35.4653137,16.0151681 L33.4888657,16.0151681 C32.0096518,16.0151681 31.9385864,16.5675073 31.9385864,17.5984881 L31.9326503,19.5768961 L35.4874341,19.5768961 L35.0684897,23.529568 L31.9357303,23.5309121 L31.9328743,35 Z"
                  />
                </g>
              </g>
            </g>
          </g>
        </svg>
        auth/sign-in-facebook
      </button>
      <button
        class="button tertiary small button google"
      >
        <svg
          class="icon icon--left"
          viewBox="0 0 22 22"
        >
          <g>
            <g
              fill="none"
              fill-rule="evenodd"
              stroke="none"
              stroke-width="1"
            >
              <g
                transform="translate(-183.000000, -256.000000)"
              >
                <g
                  transform="translate(183.000000, 256.000000)"
                >
                  <g>
                    <path
                      d="M21.56,11.25 C21.56,10.47 21.49,9.72 21.36,9 L11,9 L11,13.255 L16.92,13.255 C16.665,14.63 15.89,15.795 14.725,16.575 L14.725,19.335 L18.28,19.335 C20.36,17.42 21.56,14.6 21.56,11.25 L21.56,11.25 Z"
                      fill="#4285F4"
                    />
                    <path
                      d="M11,22 C13.97,22 16.46,21.015 18.28,19.335 L14.725,16.575 C13.74,17.235 12.48,17.625 11,17.625 C8.135,17.625 5.71,15.69 4.845,13.09 L1.17,13.09 L1.17,15.94 C2.98,19.535 6.7,22 11,22 L11,22 Z"
                      fill="#34A853"
                    />
                    <path
                      d="M4.845,13.09 C4.625,12.43 4.5,11.725 4.5,11 C4.5,10.275 4.625,9.57 4.845,8.91 L4.845,6.06 L1.17,6.06 C0.425,7.545 0,9.225 0,11 C0,12.775 0.425,14.455 1.17,15.94 L4.845,13.09 L4.845,13.09 Z"
                      fill="#FBBC05"
                    />
                    <path
                      d="M11,4.375 C12.615,4.375 14.065,4.93 15.205,6.02 L18.36,2.865 C16.455,1.09 13.965,0 11,0 C6.7,0 2.98,2.465 1.17,6.06 L4.845,8.91 C5.71,6.31 8.135,4.375 11,4.375 L11,4.375 Z"
                      fill="#EA4335"
                    />
                  </g>
                </g>
              </g>
            </g>
          </g>
        </svg>
        auth/sign-in-google
      </button>
      <div
        class="divider"
      >
        <span
          class="text"
        >
          auth/or
        </span>
      </div>
      
      <form>
        <div
          class="input-area"
        >
          <div
            class="container"
          >
            <div
              class="content"
            >
              <input
                aria-invalid="false"
                aria-label="email"
                class="input"
                id="email"
                name="email"
                type="email"
                value=""
              />
              <div
                class="placeholder"
              >
                <label
                  for="email"
                >
                  email
                </label>
              </div>
            </div>
          </div>
        </div>
        <div
          class="input-area"
        >
          <div
            class="container"
          >
            <div
              class="content"
            >
              <input
                aria-invalid="false"
                aria-label="password"
                class="input"
                id="password"
                name="password"
                type="password"
                value=""
              />
              <div
                class="placeholder"
              >
                <label
                  for="password"
                >
                  password
                </label>
              </div>
            </div>
          </div>
        </div>
        <div
          class="input-area"
        >
          <div
            class="link forgot-password"
          >
            auth/forgot-password
            ?
          </div>
        </div>
        <div
          class="input-area"
        >
          <button
            class="button primary regular button"
          >
            log-in
          </button>
        </div>
        <div
          id="recaptcha"
        />
      </form>
      <div
        class="link create-account"
      >
        auth/not-registered-yet
        ?Â 
        auth/create-account
      </div>
    </div>
  </div>
</div>
`;
